### ISSUE TITLE
Training freezes when multiple GPUs are used

### ISSUE THREAD

**Issue by rimb05** [ZFTurbo/Music-Source-Separation-Training#42]
When I try training with a single GPU (RTX 6000 Ada), it runs fine, but if I try it with multiple GPUs (using the same config file), it gets stuck here:

Instruments: ['vocal', 'guitar', 'piano', 'drums', 'bass', 'synth']
Use augmentation for training
Loading songs data from cache: /workspace/results2/metadata_1.pkl. If you updated dataset remove metadata_1.pkl before training!
Found tracks in dataset: 15300
Use multi GPU: [0, 1, 2]
Patience: 2 Reduce factor: 0.95 Batch size: 12 Grad accum steps: 1 Effective batch size: 12
Train for: 1000
Train epoch: 0 Learning rate: 9e-05
  0%|                                                  | 0/1000 [00:00<?, ?it/s]
  
  
I am calling the script like this:
!python Music-Source-Separation-Training/train.py \
    --model_type htdemucs \
    --config_path /workspace/config.yaml \
    --results_path /workspace/results/ \
    --data_path /workspace/StemGMD/training_output \
    --valid_path /workspace/StemGMD/training_output_eval \
    --dataset_type 1 \
    --num_workers 8 \
    --device_ids 0 1 2  
  
My config file:

audio:
  chunk_size: 485100 # samplerate * segment
  min_mean_abs: 0.000
  hop_length: 1024

training:
  batch_size: 4
  gradient_accumulation_steps: 1
  grad_clip: 0
  segment: 11
  shift: 1
  samplerate: 44100
  channels: 2
  normalize: true
  instruments: ['vocal', 'guitar', 'piano', 'drums', 'bass', 'synth']
  target_instrument: null
  num_epochs: 1000
  num_steps: 1000
  optimizer: adam
  lr: 9.0e-05
  patience: 2
  reduce_factor: 0.95
  q: 0.95
  coarse_loss_clip: true
  ema_momentum: 0.999
  other_fix: false # it's needed for checking on multisong dataset if other is actually instrumental
  use_amp: true # enable or disable usage of mixed precision (float16) - usually it must be true

augmentations:
  enable: true # enable or disable all augmentations (to fast disable if needed)
  loudness: true # randomly change loudness of each stem on the range (loudness_min; loudness_max)
  loudness_min: 0.5
  loudness_max: 1.5

inference:
  num_overlap: 4
  batch_size: 8

model: hdemucs

hdemucs:  # see demucs/hdemucs.py for a detailed description
  channels: 48
  channels_time: null
  growth: 2
  nfft: 4096
  wiener_iters: 0
  end_iters: 0
  wiener_residual: False
  cac: True
  depth: 6
  rewrite: True
  hybrid: True
  hybrid_old: False
  multi_freqs: []
  multi_freqs_depth: 3
  freq_emb: 0.2
  emb_scale: 10
  emb_smooth: True
  kernel_size: 8
  stride: 4
  time_stride: 2
  context: 1
  context_enc: 0
  norm_starts: 4
  norm_groups: 4
  dconv_mode: 1
  dconv_depth: 2
  dconv_comp: 4
  dconv_attn: 4
  dconv_lstm: 4
  dconv_init: 0.001
  rescale: 0.1

**Comment by rimb05** [comment]
Thanks.  The GPU memory is showing 1990MiB /  49140MiB and all three GPUs are at 100%.  I even tried batch size of 2 with the same issue.  If I only use one GPU, it works fine..

**Comment by ZFTurbo** [comment]
Can you set `num_steps: 10` and wait a bit. May be problem with progress bar...

**Comment by rimb05** [comment]
OK, I just tried torch 2.3.1 and have the same problem.  I also tried changing num_steps: 10. Do I need to delete any config files, like metadata_1.pkl when switching from single GPU to multiple GPUs?

**Comment by ZFTurbo** [comment]
> OK, I just tried torch 2.3.1 and have the same problem. I also tried changing num_steps: 10. Do I need to delete any config files, like metadata_1.pkl when switching from single GPU to multiple GPUs?

No, metadata can be the same. I don't really know what the problem is (

**Comment by rimb05** [comment]
I added some debugging and it seems to hang during the forward pass on this line:

        with torch.cuda.amp.autocast(enabled=use_amp):
               ..
                    y_ = model(x)
                
Any ideas what else I should try?  I really appreciate the help!

**Comment by ZFTurbo** [comment]
May be try to disable AMP. But you will need to decrease batch size 2 times.

**Comment by ZFTurbo** [comment]
`y_ = model(x)` - it's standard forward pass.

**Comment by rimb05** [comment]
Thanks, I tried with AMP enabled and disabled, same thing.  What I am wondering is why it works with one GPU (--device_ids 0).  If I try --device_ids 0 1 it freezes in this forward pass spot.  Even with batch size 2..

**Comment by ZFTurbo** [comment]
Can you try to choose different GPUs to work on. One by one. Put this on the top of script:

```
if __name__ == '__main__':
    import os

    gpu_use = "1,2,3"
    print('GPU use: {}'.format(gpu_use))
    os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
    os.environ["CUDA_VISIBLE_DEVICES"] = "{}".format(gpu_use)
```

May be one of GPUs doesn't work corectly?

**Comment by rimb05** [comment]
Thanks, it is solved.  I tried using a different GPU model and it works fine now.  Maybe there was some CUDA issue.  I was previously using RTX 6000 Ada, now I am using RTX A6000 and it works..  I appreciate your help with this!

**Comment by rimb05** [comment]
I have another question, if you don't mind. I switched to a different GPU, now I'm using an H100.  When I use 4 GPUs (--device_ids 0 1 2 3), the iteration speed for training is slower than if I use only one GPU.  I didn't change the batch size, it is 8.  I also tried a smaller batch size (4) and it's the same pattern..  

Speeds:

1 GPU: 1.2s/it
4 GPU:  4.5s/it

I thought that using more GPUs would make the training faster?  Am I missing another parameter?  Sorry, I am new to training with multiple GPUs..
